harness/determined/launch/autodeepspeed.py

```diff
@@ -0,0 +1,240 @@
+"""
+autodeepspeed.py is the launch layer for DeepSpeedTrial in Determined.
+
+It launches the entrypoint script using DeepSpeed's launch process.
+"""
+import argparse
+import logging
+import os
+import pathlib
+import socket
+import subprocess
+import sys
+import time
+from typing import Dict, List, cast
+
+from deepspeed.launcher.runner import DEEPSPEED_ENVIRONMENT_NAME
+
+import determined as det
+from determined import constants
+from determined.common import api
+from determined.common.api import certs
+
+
+def create_hostlist_file(
+    hostfile_path: pathlib.Path, num_proc_per_machine: int, ip_addresses: List[str]
+) -> str:
+    trial_runner_hosts = ip_addresses.copy()
+    if len(ip_addresses) == 1:
+        trial_runner_hosts[0] = "localhost"
+
+    os.makedirs(hostfile_path.parent, exist_ok=True)
+    with open(hostfile_path, "w") as hostfile:
+        lines = [f"{host} slots={num_proc_per_machine}\n" for host in trial_runner_hosts]
+        hostfile.writelines(lines)
+    return trial_runner_hosts[0]
+
+
+def create_deepspeed_env_file() -> None:
+    """Create an env var export file to pass Determined vars to the deepspeed launcher.
+
+    By default, the deepspeed launcher only keeps env vars that start with one of the following
+    ["NCCL", "PYTHON", "MV2", "UCX"].
+
+    There are certain variables that we need to be set that we can pass to deepspeed using
+    a custom env vars file.
+    """
+    INCLUDE = ["PATH", "USE_DEEPSPEED", "DET_CHIEF_IP"]
```

Comment:
why does the deepspeed launcher need USE_DEEPSPEED?

Reply:
deepspeed filters out most environment vars so we need to make sure we explicitly add back environment variables that we require.  USE_DEEPSPEED is needed on all ranks so that the distributed context is created correctly.  this differs from horovod, where we rely on the chief context being passed to all other workers so it's sufficient to just set it on the chief.

Reply:
This just means that the launcher will pass the USE_DEEPSPEED variable into the worker environments, I think.